[WIP] SourceMorph update + tests (GSoC)

TommyClausner opened the pull request:
This PR is about updating the SourceMorph class to morph SourceEstimate.

To do:
- [ ] account for mixed spaces
- [ ] add version + "See Also" + further info
- [x] add repr
- [x] remove kwargs
- [x] add tests
- [x] refactor duplication between source_estimate.py and morph.py
- [x] update what's new (before merging)
- [x] making testing correct
- [ ] think about deprecating morph wrapper in ``source_estimate.py``
- [x] check test coverage > 90% in morph.py

TommyClausner commented:
Hey together,

I hope the suggestion are implemented to your satisfaction ;)

The testing is still off I guess, how would I make it right? I'll add my thoughts when I have completed my blog post for tomorrow (will be tomorrow during the day)

Have a good new week! :)

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+    Volume Source Estimate parameters
+    ---------------------------------
+    niter_affine : tuple of int
+        Number of levels (``niter_affine.__len__()``) and number per level of
```
you men `len(n_iter_affine)` ?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+        for the tuple mean later levels and each int represents the number
+        of iterations in that level. Default is niter_affine=(100, 100, 10)
+    niter_sdr : tuple of int
+        Number of levels (``niter_sdr.__len__()``) and number per level of
```
same comment here

jasmainak replied:
also I would decompress the docstring a bit. For example, even after reading the docstring twice, I am not sure what is a "level"

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+        vertices to morph to. If morphing a volume source space, subject_to can
+        be the path to a MRI volume.
+    niter_affine : tuple of int
+        Number of levels (``niter_affine.__len__()``) and number per level of
```
same

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+        of iterations in that level. Default is niter_affine=(100, 100, 10)
+    niter_sdr : tuple of int
+        Number of levels (``niter_sdr.__len__()``) and number per level of
+        iterations to perform the sdr transform. Increasing index values
```
what is "sdr" ?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+        for the tuple mean later levels and each int represents the number
+        of iterations in that level.
+    niter_sdr : tuple of int
+        Number of levels (``niter_sdr.__len__()``) and number per level of
```
len(...)

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+                               'subject_from')
+        self.morph_data = _compute_morph_data(self, src, verbose=verbose)
+
+    def __call__(self, stc_from, verbose=None):
```
verbose decorator is not needed here?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+                             'morph.subject_from must match')
+
+        # if VolSourceEstimate update mri properties for '.as_volume'
+        if self.kind is 'volume':
```
nopes. Do

```py
self.kind == 'volume'
```

see: https://stackoverflow.com/questions/1504717/why-does-comparing-strings-in-python-using-either-or-is-sometimes-produce/1504742#1504742

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+            s = "%s" % self.kind
+            s += ", subject_from : %s" % self.subject_from
+            s += ", subject_to : %s" % self.subject_to
+            if self.kind == 'volume':
```
does it need to be nested?

```py
>>> python -c 'import this' | sed '7!d'
Flat is better than nested.
```

jasmainak commented on examples/inverse/plot_use_sourcemorph.py:
```diff
@@ -0,0 +1,101 @@
+a common reference space. It will be demonstrated using the SourceMorp class.
+The example uses parts of the MNE example
+:ref:`sphx_glr_auto_examples_inverse_plot_lcmv_beamformer_volume.py` and
+:ref:`sphx_glr_auto_examples_inverse_plot_lcmv_beamformer.py`.
```
Is this true anymore?

jasmainak commented on examples/inverse/plot_use_sourcemorph.py:
```diff
@@ -0,0 +1,101 @@
+case and using an affine transform in the SourceEstimate or
+VectorSourceEstimate case. Afterwards the transformation will be applied to the
+beamformer result. The result will be a plot showing the morphed result
+overlaying the fsaverage T1. Uncomment at the respective location to plot the
```
which line to uncomment? I'd remove this from here and just write it where you have to uncomment

jasmainak commented on examples/inverse/plot_use_sourcemorph.py:
```diff
@@ -0,0 +1,101 @@
+
+# Save and load SourceMorph
+# source_morph.save('vol')
+# source_morph = mne.read_source_morph('vol-morph.h5')
```
move to tests

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+                             'path to a saved instance of SourceSpaces')
+
+        if subject_from is None:
+            try:
```
you can simply do:

```py
`if not 'subject_his_id'` in src[0]:
   raise KeyError('...')
```

jasmainak replied:
then you don't also need `subject_from`

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+
+        if isinstance(src, SourceSpaces):
+            self.kind = src.kind
+            self.src = src
```
why do you assign it if `self.src` is never used later?

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+        self.xhemi = xhemi
+        self.morph_data = None
+
+        if src is None:
```
so it will fail silently?

TommyClausner replied:
necessary to create an "empty" object for loading

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,731 @@
+            if mri_resolution:
+                new_zooms = self.morph_data['mri_zooms']
+            else:
+                new_zooms = None
```
avoid nesting when you can :) Here you can do:

```py
new_zooms = None
if mri_resolution:
   new_zooms = self.morph_data['mri_zooms']
```

Think about the person reading the code. The fewer lines of code and simpler the code, the easier it is for the code to be maintained

TommyClausner replied:
and what if mri_resolution can also be a tuple (which it can)?
I do now:
```
new_zooms = None

        if isinstance(mri_resolution, bool) and mri_resolution:
            new_zooms = self.morph_data['mri_zooms']

        elif isinstance(mri_resolution, tuple):
            new_zooms = mri_resolution
```

jasmainak commented on mne/tests/test_source_morph.py:
```diff
@@ -0,0 +1,93 @@
+from mne.minimum_norm import apply_inverse, read_inverse_operator
+from mne.utils import run_tests_if_main, requires_nibabel
+
+print(__doc__)
```
we don't do this in tests

jasmainak commented on mne/tests/test_source_morph.py:
```diff
@@ -0,0 +1,93 @@
+    assert isinstance(stc_surf_morphed, SourceEstimate)
+
+
+@requires_nibabel()
```
does not require dipy?

jasmainak commented on mne/tests/test_source_morph.py:
```diff
@@ -0,0 +1,93 @@
+    source_morph_vol.save('vol')
+    source_morph_vol.save('vol.h5')
+
+    source_morph_vol = read_source_morph('vol.h5')
```
you should check if the instance is the same after reading ...

jasmainak commented on mne/tests/test_source_morph.py:
```diff
@@ -0,0 +1,93 @@
+    source_morph_vol = read_source_morph('vol.h5')
+    stc_vol_morphed = source_morph_vol(stc_vol)
+    os.remove('vol-morph.h5')
+    os.remove('vol.h5')
```
use a tempdir. See how we do it in other tests

jasmainak commented on mne/tests/test_source_morph.py:
```diff
@@ -0,0 +1,93 @@
+    source_morph_surf = read_source_morph('surf.h5')
+    stc_surf_morphed = source_morph_surf(stc_surf)
+    os.remove('surf-morph.h5')
+    os.remove('surf.h5')
```
again, see how we do the tests. It should be in a tempdir

agramfort commented:
@TommyClausner start early with the writing of the test.

FYI you can use coverage as quantified measure of testing progress. For example:

`pytest --cov=mne.source_estimate --cov-report="term" mne/tests/test_source_estimate.py -k test_morph_data``

to only run one test `test_morph_data` and see the coverage report in the terminal.

also keep in mind that for surfaces it's not a new feature so you need to integrate all this with what already exists.

TommyClausner commented:
Al right, updates implemented.

Some remarks:
@agramfort I wouldn't like to move the Surface + Vector, because I actually don't test the functionality with the other functions. Can this test file not be just class related?

@jasmainak for the mri_resolution=bool : the problem is indeed, that it also can be a tuple, but I tried to make it not nested if statements now whenever somewhat reasonably possible ;)
For the "silent fail" - it's just to return an empty object, but with all the attributes set up

I had a coverage of 95% before it stopped working (still don't know why, seems to be a common problem: ``Exception TypeError: "'NoneType' object is not callable" in [...] tmp_mne_tempdir_Iptsps'> ignored``), but I guess it should be slightly higher now, because the only two things I'm not testing are: full-res morph (computation time) and not having nibabel or dipy (son't know how it would be best(?) ;) )

agramfort commented:
@TommyClausner 2 things:
- do not open a PR from the master branch. I cannot push your branch to help now.
- SourceEstimate.morph should now use the same code.

presently a lot of functionality/code is duplicated between morph functions in source_estimate.py and in morph.py

you need to fix this ASAP.

jasmainak commented:
@TommyClausner let us know when we can review again

jasmainak commented on doc/whats_new.rst:
```diff
@@ -19,6 +19,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Add :class:`mne.SourceMorph` with :meth:`mne.SourceMorph.save` and :meth:`mne.SourceMorph.as_volume`, :func:`mne.read_source_morph`, :func:`mne.read_source_morph`, example plot_use_source_morph.py, by `Tommy Clausner`
```
normally, examples are not listed here. Also, you need to add your webpage at the bottom for the link to render properly

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,715 @@
+            mri_path_to = static_path
+        else:
+            mri_path_to = op.join(static_path, mri_subpath)
+        try:
```
avoid try-catch pattern when you can. Here, for instance you can check if the file exists using `op.isfile` ...

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,715 @@
+
+        data_to = None
+        if morph.subject_to == 'fsaverage':
+            data_to = [np.arange(10242)] * len(data_from)
```
why is this number hard coded? where does it come from?

jasmainak commented on examples/inverse/plot_use_sourcemorph.py:
```diff
@@ -0,0 +1,92 @@
+                           subject_from='sample',
+                           subject_to='fsaverage',
+                           subjects_dir=subjects_dir,
+                           grid_spacing=(5., 5., 5.))
```
let's remove this line since we have it as default now. Keep it simple stupid for the user :)

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,715 @@
+        If True, warn if not all vertices were used.
+    xhemi : bool
+        Morph across hemisphere. Currently only implemented for
+        ``subject_to == subject_from``. See notes below.
```
I don't see anything about this in the Notes section

jasmainak commented on mne/morph.py:
```diff
@@ -0,0 +1,715 @@
+                   overwrite=True)
+        logger.info('[done]')
+
+    def as_volume(self, stc, fname=None, mri_resolution=False):
```
this needs to be refactored right?

jasmainak commented:
@TommyClausner red lines are more appreciated than green lines on github :) Let us know once you have refactored the code to avoid duplication

TommyClausner commented:
sorry didn't see your comments - I'll push again  in some minutes ;)

TommyClausner commented:
all right some comments:
I now moved all functionality that is morph related from ``source_morph.py`` to ``morph.py``, further I tried to assure that imports from ``.source_estimate`` were changed (only for viz/_3d.py). Morph tests are moved from ``test_source_estimate.py`` to ``test_morph.py``

Further I added deprecation decorators to all source_estimate morphing functions, but not ``as_volume`` - However ``buffer_size`` and ``n_jobs`` are useless now and I added a statement to the description.

Also: I wasn't entirely finished  before I was forced to push ( :P ) so there might still be stuff in there, that doesn't make sense or has the wrong name or so...

TommyClausner commented:
ok I pushed it mostly to see what CircleCI does and if all examples etc are correct and the change didn't f up something... don't read it yet ;)

jasmainak commented:
@TommyClausner I would have just used relative imports instead of moving everything. Specially since the functions in `source_estimate.py` are getting deprecated anyway? Creates a huge diff that is difficult to read ... 

also for the future can you can follow the [numpy convention for commit messages](https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message)

jasmainak commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,438 @@
+    assert_raises(ValueError, stc.as_volume, src, mri_resolution=(4., 4., 4.))
+
+
+@pytest.mark.slowtest
```
how long do the tests take?

TommyClausner replied:
my bench says:
29.64s call     mne/tests/test_morph.py::test_volume_source_morph
13.96s call     mne/tests/test_morph.py::test_morph_data
2.70s call     mne/tests/test_morph.py::test_surface_vector_source_morph
0.44s call     mne/tests/test_morph.py::test_save_vol_stc_as_nifti
0.39s call     mne/tests/test_morph.py::test_stc_as_volume

jasmainak replied:
okay you might want to bring it down to 10 seconds. 30 seconds is huge !

TommyClausner replied:
hm... so splitting it into sub-tests?

jasmainak replied:
haha no. I mean actually making it faster. Maybe try profiling your code and seeing what takes time. Try to use fewer channels / source points. Decimated grid, less iterations. There are different tricks you can use to make things faster

jasmainak replied:
maybe see what you can precompute and load the precomputed file.

TommyClausner replied:
the long lasting stuff is the computation of the non-linear morph... I see how much I can strip it - otherwise I already load everything

TommyClausner replied:
guess that's all I can offer:
13.82s call     mne/tests/test_morph.py::test_morph_data
10.08s call     mne/tests/test_morph.py::test_volume_source_morph
2.62s call     mne/tests/test_morph.py::test_surface_vector_source_morph
0.47s call     mne/tests/test_morph.py::test_save_vol_stc_as_nifti
0.38s call     mne/tests/test_morph.py::test_stc_as_volume

jasmainak commented on mne/tests/test_morph.py:
```diff
@@ -0,0 +1,438 @@
+
+
+@requires_nibabel()
+@pytest.mark.slowtest
```
same question here

TommyClausner commented:
@jasmainak actually we discussed this on Hangouts (sorry for the exclusion) - in general it's probably good practice to move the morphing functionality into ``morph.py`` - mostly because the morphing in ``source_estimate.py`` will be deprecated. So there is no need for it to stay there. And for the ``as_volume``: as far as I understood it will not be deprecated in ``source_estimate.py`` but it's necessary to have it on ``morph.py`` in order to output a volume correctly after morphing - so I guess importing from ``morph`` into ``source_estimate`` makes sense after all, no?

agramfort commented:
much better

I will look at the diff ASAP.

make CIs happy

TommyClausner commented:
Hey err... does anyone know how I can solve the ``ImportError`` when using ``assert_raises``? before I imported it from ``nose`` directly, which made it crash and now I went via ``numpy`` and it still crashes :/ - any input on that would be really appreciated :D

agramfort commented:
point us to the problematic line / error.

larsoner commented:
Use pytest.raises

>

TommyClausner commented on mne/source_estimate.py:
```diff
@@ -2275,9 +2272,6 @@ def morph_data(subject_from, subject_to, stc_from, grade=5, smooth=None,
 
-@deprecated("This function is deprecated and might be removed in a future "
-            "release. Use morph = SourceMorph(src_from) and "
-            "stc_to = morph(stc_from)")
```
yeah I also found it strange, but it was definitely causing it, since I also tried on different functions and it caused the same...

larsoner replied:
Did you also deprecate `stc.morph` and `stc.morph_precomputed`? You should deprecate these as well. That might be the problem.

If all else fails you can add them to the ignores list in `test_docstring.py`, but this shouldn't really be necessary

TommyClausner replied:
tried, then I get even more errors...

TommyClausner replied:
ok weird, I can deprecate ``stc.morph_precomputed`` but not ``stc.morph``
